src/components/Heading/Heading.stories.mdx

```diff
@@ -78,7 +78,9 @@ semantic elements can be overwritten using the `displaySize` prop.
     {(args) => (
       <Heading
         {...args}
-        displaySize={sizesEnumValues.getValue(args.displaySize)}
+        displaySize={
+          args.displaySize && sizesEnumValues.getValue(args.displaySize)
```

Comment:
This was the main issue. `displaySize` was always set to a value instead of the default undefined and it overrode the display size from the initial `level` value. So now, if it _should_ be undefined, then it will be.
All the other changes are related and more like refactoring.